src/update-index.js

```diff
@@ -43,37 +53,81 @@ const error = (path, e) => ({
 /**
  * Find an existing entry, given its path or sourceHash and branch.
  *
- * @param {object} index Algolia index
+ * @param {SearchIndex} index Algolia index
  * @param {object} attributes Attributes to search for
- * @param {string} branch Branch
  */
 async function search(index, attributes) {
-  const filters = Object.getOwnPropertyNames(attributes).map(
-    (name) => `${name}:${attributes[name]}`,
-  );
+  const filters = Object.getOwnPropertyNames(attributes)
+    .filter(
+      (name) => attributes[name],
+    )
+    .map(
+      (name) => `${name}:${attributes[name]}`,
+    );
   const searchresult = await index.search({
     attributesToRetrieve: ['path', 'name', 'objectID', 'sourceHash'],
     filters: filters.join(' AND '),
   });
   return searchresult.nbHits !== 0 ? searchresult.hits[0] : null;
 }
 
+/**
+ * Prepare items to be re-indexed.
+ *
+ * @param {object} cfg index configuration
+ * @param {SearchIndex} index algolia index
+ * @param {object} coll collection of items
+ * @param {string} branch GitHub branch
+ *
+ * @return {Array} of { path, hit } items
+ */
+async function prepareItems(cfg, index, coll, branch) {
+  let searchresults = (await Promise.all(coll.items.map(
+    // try to lookup path for items that only have a source hash
+    async (item) => {
+      const { path, sourceHash } = item;
+      const hit = await search(index, { path, sourceHash, branch });
+      return { path: path || (hit ? hit.path : null), hit };
+    },
+  ))).filter(
+    // drop tems that still haven't a valid path
+    ({ path }) => !!path,
+  );
+  const { mountpoint } = coll;
+  if (mountpoint) {
+    // if mountpoint is given, translate paths and remove leading slash
+    const re = new RegExp(`^${mountpoint.root}/`);
+    const repl = mountpoint.path.replace(/^\/+(.*)/, '$1');
+
+    searchresults = searchresults.map(
+      (item) => ({ path: item.path.replace(re, repl), hit: item.hit }),
+    );
+  }
+  return searchresults.filter(
+    // keep only items that are included in the index definition
+    ({ path }) => includes(index, path),
+  ).map(({ path, hit }) => {
+    // replace requested extension with the one in source
+    const noext = path.replace(/([^.]+)\.[^./]+/, '$1');
```

Comment:
I think this is more complicated. for example, onedrive you have `.docx` documents, but google docs, don't have extentions....

also, I think the regexp is wrong for the case where path segment contains a dot, eg: `/ui.apps/content.md`.

Reply:
OK, then again, we need a formal definition of what google docs (or others look like). I know I repeat myself, but my change was only meant for `onedrive`...

I don't think that `\.[^./]+` matches anything but the last dot after a slash, but I could be mistaken.

Reply:
![image](https://user-images.githubusercontent.com/917628/76757061-43fe1080-67ca-11ea-9294-1aa804e8352f.png)

Reply:
Oops, right, the end-of-string `$` was missing: `/([^.]+)\.[^./]+$/`

Reply:
> Oops, right, the end-of-string `$` was missing: `/([^.]+)\.[^./]+$/`

but do you think it's more performant than doing a `lastIndexOf` ?

Reply:
Nope, this was just out of brevity :) To be fair, you'd probably need two of those `lastIndexOf`, though, right?
```
const lastSlash = path.lastIndexOf('/')`
if (lastSlash === -1) {
  lastSlash = path.length();
}
const ext = lastIndexOf('.', lastSlash)
...
```

Reply:
https://github.com/adobe/helix-pipeline/blob/974a921fd884e0589fb29a99bd3394133ed84a95/src/html/fetch-external.js#L38-L40

Reply:
Hm, but this would also remove a `selector`, isn't that a problem?

Reply:
you're right... but in that case, the `resource path` doesn't have a selector.

so maybe it should also be the provider that produces the resource path... eg:

| type | item path             | resourcepath      |
|:-----|:----------------------|:------------------|
| 1d   | /docs/foo.docx        | /docs/foo         |
| 1d   | /docs/2020.goals.docx | /docs/2020.goals  |
| gh   | /content/en/index.md  | /content/en/index |
| goo  | /site/My.Document     | /site/My.Docuemnt |

Reply:
My current proposal stores the `domain` of observed changes in `helix-query.yaml` (see https://github.com/adobe/helix-index-files/issues/89). The check whether a path is included in that domain is currently performed in `helix-index-files`, but if we strip the extension already in the provider, changes to a `1d` file named `foo.gif` would be transferred as `foo` and would eventually try to request `foo.html` in `helix-index-files`, which looks wrong.

So we either need to:
- keep the extension around, may be transmit it in a separate property?
- filter the paths already in the listener, but this would leak knowledge of `helix-query.yaml` to the listener, which currently only looks at `fstab.yaml`
- hardcode the list of extensions observed (`md` + `docx` for `1d`, etc), but this sounds wrong again

Reply:
but maybe the include/exclude is applied to the wrong domain.... (I'll comment in #89)